tiledb/ml/data_apis/tensorflow_sparse.py

```diff
@@ -0,0 +1,158 @@
+        of the data. For example, in case the tile extent of the first dimension of a TileDB array (x or y) is equal to
+        32, it's recommended to set batch_size=32. Any batch size will work, but in case it's not equal the tile extent
+        of the first dimension of the TileDB array, you won't achieve highest read speed. For more details on tiles,
+        tile extent and indices in TileDB, please check here:
+        https://docs.tiledb.com/main/solutions/tiledb-embedded/performance-tips/choosing-tiling-and-cell-layout#dense-arrays
+        """
+        cls.x = x_array
+        cls.y = y_array
+
+        # Check that x and y have the same number of rows
+        if x_array.schema.domain.shape[0] != y_array.schema.domain.shape[0]:
+            raise Exception(
+                "X and Y should have the same number of rows, i.e., the 1st dimension "
+                "of TileDB arrays X, Y should be of equal domain extent."
+            )
+
+        # Get number of observations
+        rows = x_array.schema.domain.shape[0]
+
+        # Get x and y shapes
+        cls.x_shape = (None,) + x_array.schema.domain.shape[1:]
+        cls.y_shape = (None,) + y_array.schema.domain.shape[1:]
+
+        # Get x and y data types
+        cls.x_dtype = x_array.schema.attr(0).dtype
+        cls.y_dtype = y_array.schema.attr(0).dtype
+
+        if isinstance(y_array, tiledb.SparseArray):
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_sparse,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.SparseTensorSpec(shape=cls.y_shape, dtype=cls.y_dtype)
+                ),
+                args=(rows, batch_size)
+            )
+        else:
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_dense,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.TensorSpec(shape=cls.y_shape, dtype=cls.y_dtype)
+                ),
+                args=(rows, batch_size)
+            )
+
+    @classmethod
+    def _generator_sparse_sparse(cls, rows: int, batch_size: int) -> tuple:
+        """
+        A generator function that yields the next training batch.
+        :param x: TileDB array. An opened TileDB array which contains features.
+        :param y: TileDB array. An opened TileDB array which contains labels.
+        :param rows: Integer. The number of observations in x, y datasets.
+        :param batch_size: Integer. Size of batch, i.e., number of rows returned per call.
+        :return: Tuple. Tuple that contains x and y batches.
+        """
+        # Loop over batches
+        # https://github.com/tensorflow/tensorflow/issues/44565
+        for offset in range(0, rows, batch_size):
+            y_batch = cls.y[offset: offset + batch_size]
```

Comment:
This only works for integer dimensions. @georgeSkoumas lets sync up on this tomorrow. The dataset api here seems to only be for the features and labels and not for the training data? I'd like to better understand the scope of this integration with tensorflow and its use cases.

Reply:
@Shelnutt2 that's true. We were thinking of supporting the basic scenario here (integer dimensions) for x and y and iterate later in case we need to support string dimensions or other more complicated scenarios. As you mentioned, X are features  and Y are labels, which constitute a training dataset for supervised learning problems, e.g, classification, regression. We can sync up tomorrow.

Reply:
In the complicated scenarios for our discussion tomorrow we could also add the approach for multiplicities of cells